### ISSUE TITLE
A bug about the `device` property in the class `BaseFrameworkExecutor`

### ISSUE THREAD

**Issue by Wei-weiWang** [jina-ai/jina#643]
I have found a bug in your `BaseFrameworkExecutor` class about the `device` property. Here is the link ([https://github.com/jina-ai/jina/blob/master/jina/executors/frameworks.py#L10](link)).

The attribute `_device` is assigned in the `post_init` method and used in the `device` property. In the `ImagePaddlehubEncoder` class, however, the `device` property is used without calling `post_init` in the first hand. Therefore, an exception will be raised. I suggest in the `device` property, the existence if the `_device` attribute shuold be checked so as to improve the robustness of the codes. Attached is the suggested revision.

```
class BaseFrameworkExecutor(BaseExecutor):
    """
    :class:`BaseFrameworkExecutor` is the base class for the executors using other frameworks internally, including
        `tensorflow`, `pytorch`, `onnx`, `faiss` and `paddlepaddle`.
    """
    def __init__(self, model_name: str = None, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.model_name = model_name
    
    def post_init(self):
        super().post_init()
        self._device = None
        if self.on_gpu:
            try:
                cuda_version = subprocess.check_output(['nvcc', '--version']).decode()
                self.logger.success(f'CUDA compiler version: {cuda_version}')
            except OSError:
                self.logger.warning('on_gpu=True, but you dont have CUDA compatible GPU, i will reset on_gpu=False ')
                self.on_gpu = False

    @property
    def device(self):
        """Set the device on which the executor will be running.
        ..notes:
            In the case of using GPUs, we only use the first gpu from the visible gpus. To specify which gpu to use,
            please use the environment variable `CUDA_VISIBLE_DEVICES`.
        """
        if hasattr(self, '_device') and self._device is not None:
            return self._device
        else:
            try:
                self._device = self.get_device()
                return self._device
            except Exception:
                self.logger.error(f'error when setting devices "on_gpu={self.on_gpu}"')
                raise

    @abstractmethod
    def get_device(self):
        pass

    def to_device(self, *args, **kwargs):
        """Put the model on specified device (``on_gpu``) and returns the device context"""
        raise NotImplementedError
```

**Comment by nan-wang** [comment]
hi, @Wei-weiWang Would you please provide more information about your jina version? We've recently made some changes recently at [https://github.com/jina-ai/jina/pull/636](https://github.com/jina-ai/jina/pull/636).

**Comment by Wei-weiWang** [comment]
Sorry for just seeing this. I use the "jinaai/jina:devel" version pulled
from the dockerhub, which is "0.3.8-devel" currently. Also, I found the
same bug in the master branch of your github repo. See
https://github.com/jina-ai/jina/blob/master/jina/executors/frameworks.py#L10
.

Nan Wang <notifications@github.com> 于2020年7月11日周六 下午10:11写道：

**Comment by JoanFM** [comment]
Hello @Wei-weiWang. Great catch! 

The proper solution seems to be to force child classes to call `super().__post_init__()` to force the setting of `_device`.

The solution is addressed in #645, and the issue should be solved.
